File: frankpiv/backend/frankx_backend.py
from abc import ABC

from frankx import Robot, Affine, LinearMotion

from frankpiv.backend.general import GeneralBackend


class Backend(GeneralBackend, ABC):

    def __init__(self, config):
        super().__init__(config)
        self.frankx_config = config["frankx"]
        self._robot = None

    def _initialize(self):
        self._robot = Robot(self.frankx_config["robot_name"], self.frankx_config["dynamic_rel"])
        self._robot.set_dynamic_rel(self.frankx_config["dynamic_rel"])

    def _finish(self):
        self._robot = None

    # TODO fix in lab
    def _current_pose(self) -> Affine:
        return self._robot.current_pose(Affine())

    def _move_robot_cartesian(self, target_pose: Affine):
        self._robot.move_cartesian(LinearMotion(target_pose))

File: src/frankx_backend.cpp
#include <utility>

#include "frankpiv/frankx_backend.hpp"
#include "frankpiv/utilities.hpp"

using namespace Eigen;

namespace frankpiv::backend {
    FrankxBackend::FrankxBackend(const YAML::Node &config, std::string node_name) : GeneralBackend(config, std::move(
            node_name)) {
        YAML::Node frankx_config = config["frankx"];
        this->fci_ip = frankx_config["fci_ip"].as<std::string>();
        this->dynamic_rel = frankx_config["dynamic_rel"].as<double>();
        this->robot = nullptr;
    }

    void FrankxBackend::initialize() {
        frankx::Robot _robot = frankx::Robot(this->fci_ip, this->dynamic_rel);
        this->robot = &_robot;
    }

    void FrankxBackend::finish() {
        this->robot = nullptr;
    }

    Eigen::Affine3d FrankxBackend::currentPose() {
        return this->robot->currentPose(frankx::Affine()).data;
    }

    void FrankxBackend::moveRobotCartesian(const Eigen::Affine3d &target_pose) {
        frankx::LinearMotion motion = frankx::LinearMotion(movex::WaypointMotion::Affine(target_pose));
        this->robot->move(motion);
    }
}